eta/core/datasets.py

```diff
@@ -1338,3 +1375,842 @@ def required(cls):
+    '''
+
+    def __init__(self, attribute_name, object_label=None,
+                 target_quantile=0.25, negative_power=5, target_count=None):
+        '''Creates a Balancer instance.
+
+        Args:
+            attribute_name (str): the name of the attribute to balance by
+            object_label (str): the name of the object label that the
+                attribute_name must be nested under. If this is None, it is
+                assumed that the attributes are Image/Frame level attrs.
+            target_quantile (float): value between [0, 1] to specify what the
+                target count per attribute value will be.
+                0.5 - will result in the true median
+                0   - the minimum value
+                It is recommended to set this somewhere between [0, 0.5]. The
+                smaller this value is, the closer all values can be balanced,
+                at the risk that if some values have particularly low number of
+                samples, they dataset will be excessively trimmed.
+            negative_power (float): value between [1, ~LARGE~] that weights the
+                negative values (where the count of a value is less than the
+                target) when computing the score for a set of indices to remove.
+                1 - will weight them the same as positive values
+                2 - will square the values
+                ...
+                Check Balancer._solution_score for more details.
+            target_count (int): override target count for each attribute value.
+                If this is provided, target_quantile is ignored.
+        '''
+        self.attr_name = attribute_name
+        self.object_label = object_label
+        self.target_quantile = target_quantile
+        self.negative_power = negative_power
+        self.target_count = target_count
+
+    def transform(self, src):
+        '''Modify the BuilderDataset records by removing records until the
+        target attribute is ~roughly~ balanced for each value.
+
+        Args:
+            src (BuilderDataset): the dataset builder
+        '''
+        # STEP 1: Get attribute value(s) for every record
+        occurrence_matrix, attribute_values, record_idxs = \
+            self._get_occurrence_matrix(src.records)
+        if not attribute_values:
+            return
+
+        # STEP 2: determine target number to remove of each attribute value
+        counts = np.sum(occurrence_matrix, axis=1).astype(np.dtype('int'))
+        target_count = self._get_target_count(counts)
+
+        # STEP 3: find the records to keep
+        keep_idxs = self._get_keep_idxs(occurrence_matrix, counts, target_count)
+
+        # STEP 4: modify the list of records
+        old_records = src.records
+        src.clear()
+        for ki in keep_idxs:
+            src.add(old_records[record_idxs[ki]])
+
+    def _get_occurrence_matrix(self, records):
+        '''Compute occurrence of each attribute value for each class
```

Comment:
FYI- our style doesn't require writing docstrings for private methods, if you don't want to